src/uct/ib/ud/base/ud_iface.c

```diff
@@ -889,19 +942,56 @@ ucs_status_t uct_ud_iface_event_arm(uct_iface_h tl_iface, unsigned events)
 
 void uct_ud_iface_progress_enable(uct_iface_h tl_iface, unsigned flags)
 {
-    uct_ud_iface_t *iface = ucs_derived_of(tl_iface, uct_ud_iface_t);
+    uct_ud_iface_t *iface       = ucs_derived_of(tl_iface, uct_ud_iface_t);
+    ucs_async_context_t *async  = iface->super.super.worker->async;
+    ucs_async_mode_t async_mode = async->mode;
+    ucs_status_t status;
+
+    uct_ud_enter(iface);
 
     if (flags & UCT_PROGRESS_RECV) {
-        uct_ud_enter(iface);
         iface->rx.available += iface->rx.quota;
         iface->rx.quota      = 0;
         /* let progress (possibly async) post the missing receives */
-        uct_ud_leave(iface);
     }
 
+    if (!iface->tx.timer_disable && (iface->async.timer_id == 0)) {
```

Comment:
do we really need to check iface->tx.timer_disable?
if iface->tx.timer_disable==1, the async handler may exist, but ep will not add their timers to the timer wheel anyway